packages/csvviewer/src/model.ts

```diff
@@ -0,0 +1,676 @@
+      // Row isn't in the cache, so we invalidate the entire cache and set up
+      // the cache to hold the requested row.
+      this._columnOffsets.fill(0xFFFFFFFF);
+      this._columnOffsetsStartingRow = row;
+      rowIndex = 0;
+    }
+
+    // Check to see if we need to fetch the row data into the cache.
+    if (this._columnOffsets[rowIndex] === 0xFFFFFFFF) {
+      // Figure out how many rows below us also need to be fetched.
+      let maxRows = 1;
+      while (maxRows <= this._maxCacheGet && this._columnOffsets[rowIndex + maxRows * ncols] === 0xFFFFFF) {
+        maxRows++;
+      }
+
+      // Parse the data to get the column offsets.
+      let {offsets} = PARSERS[this._parser]({
+        data: this._data,
+        delimiter: this._delimiter,
+        rowDelimiter: this._rowDelimiter,
+        quote: this._quote,
+        columnOffsets: true,
+        maxRows: maxRows,
+        ncols: ncols,
+        startIndex: this._rowOffsets[row]
+      });
+
+      // Copy results to the cache.
+      for (let i = 0; i < offsets.length; i++) {
+        this._columnOffsets[rowIndex + i] = offsets[i];
+      }
+    }
+
+    // Return the offset index from cache.
+    return this._columnOffsets[rowIndex + column];
+  }
+
+  /**
+   * Parse the data string asynchronously.
+   *
+   * #### Notes
+   * It can take several seconds to parse a several hundred megabyte string, so
+   * we parse the first 500 rows to get something up on the screen, then we
+   * parse the full data string asynchronously.
+   */
+  private _parseAsync(): void {
+    // Number of rows to get initially.
+    let currentRows = this._initialRows;
+
+    // Number of rows to get in each chunk thereafter. We set this high to just
+    // get the rest of the rows for now.
+    let chunkRows = Math.pow(2, 32) - 1;
+
+    // We give the UI a chance to draw by delaying the chunk parsing.
+    let delay = 30; // milliseconds
+
+    // Define a function to parse a chunk up to and including endRow.
+    let parseChunk = (endRow: number) => {
+      try {
+        this._computeRowOffsets(endRow);
+      } catch (e) {
+        // Sometimes the data string cannot be parsed with the full parser (for
+        // example, we may have the wrong delimiter). In these cases, fall back to
+        // the simpler parser so we can show something.
+        if (this._parser === 'quotes') {
+          console.warn(e);
+          this._parser = 'noquotes';
+          this._resetParser();
+          this._computeRowOffsets(endRow);
+        } else {
+          throw e;
+        }
+      }
+      return this._doneParsing;
+    };
+
+    // Reset the parser to its initial state.
+    this._resetParser();
+
+    // Parse the first rows to give us the start of the data right away.
+    let done = parseChunk(currentRows);
+
+    // If we are done, return early.
+    if (done) {
+      return;
+    }
+
+    // Define a function to recursively parse the next chunk after a delay.
+    let that = this;
```

Comment:
Can this function not be done as an arrow function without needing `that`?
```typescript
// Define a function to recursively parse the next chunk after a delay.
let delayedParse = () => {
  // Parse up to the new end row.
  let done = parseChunk(currentRows + chunkRows);
  currentRows += chunkRows;

  // Gradually double the chunk size until we reach a million rows, if we
  // start below a million-row chunk size.
  if (chunkRows < 1000000) {
    chunkRows *= 2;
  }

  // If we aren't done, the schedule another parse.
  if (done) {
    this._delayedParse = null;
  } else {
    this._delayedParse = window.setTimeout(delayedParse, delay);
  }
}
```

Reply:
You do the same above, so I assume there's a reason I am missing.

Reply:
It's the recursion. I need a name for the function so I can call it to set the timeout.

Reply:
huh, will what you wrote work? I'm not sure if the closure will include the right variable. I guess it's easy to check...

I just checked - it seems to work okay in Firefox.

Reply:
And same in Chrome: 

```javascript
let x = () => {console.log(setTimeout(x, 1000))}
x()
```
Okay, commit coming up.